### ISSUE TITLE
@ExtendWith on instance fields registered only after test instantiation

### ISSUE THREAD

**Issue by kostislav** [junit-team/junit5#2847]
### Overview

JUnit version: 5.8.2

Issue #864 is awesome. However, when I tried to use it, I hit a problem and I'm not sure if (or why) it is supposed to work like this. For some reason, non-static fields are only checked for extensions _after_ the test instance is created, so it is impossible to register a `ParameterResolver` extension that would resolve the value of a private final field.

This behavior makes sense for `@RegisterExtension` fields, as it needs the instance to be registered and that is only available after creating the test instance. It does not, however, make sense for `@ExtendWith` annotations, as those can be read even without having an actual instance.

```java
@lombok.RequiredArgsConstructor
class TestClass {
  @TestAnnotation
  private final String someField;

  @Test
  void someTestMethod() {
    assertThat(someField, is("abc"));
  }
}

@ExtendWith(TestAnnotationExtension.class)
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.FIELD)
@interface TestAnnotation {
}

class TestAnnotationExtension implements ParameterResolver {
  @Override
  public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
    return parameterContext.getParameter().getType().equals(String.class);
  }

  @Override
  public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
    return "abc";
  }
}

```

### Expected outcome:

test passes.

### Actual outcome:

```
org.junit.jupiter.api.extension.ParameterResolutionException: No ParameterResolver registered for parameter [final java.lang.String someField] in constructor [public TestClass(java.lang.String)]
```

**Comment by sbrannen** [comment]
I've edited your comment to improve the formatting. You might want to check out this [Mastering Markdown guide](https://guides.github.com/features/mastering-markdown/) for future reference.

**Comment by sbrannen** [comment]
> Issue #864 is awesome. 

Glad you like it!

> so it is impossible to register a `ParameterResolver` extension that would resolve the value of a private final field.

A `ParameterResolver` does not resolve fields. It resolves parameters. Based on your example, I assume you meant "it is impossible to register a `ParameterResolver` extension that would resolve the value for a constructor parameter".

> It does not, however, make sense for `@ExtendWith` annotations, as those can be read even without having an actual instance.

It's true that the instance is not needed for `@ExtendWith` on fields. We'll investigate if it's feasible to change the logic here.

**Comment by sbrannen** [comment]
FYI: your use case is actually already supported if you annotate the constructor parameter instead of the field.

For example, the following test class passes.

```java
public class TestClass {

	private final String someField;

	TestClass(@TestAnnotation String someField) {
		this.someField = someField;
	}

	@Test
	void someTestMethod() {
		assertThat(someField).isEqualTo("abc");
	}
}

@Target({ ElementType.PARAMETER, ElementType.FIELD })
@Retention(RetentionPolicy.RUNTIME)
@ExtendWith(TestAnnotationExtension.class)
@interface TestAnnotation {
}
```

**Comment by kostislav** [comment]
@sbrannen Thank you and sorry for the improper formatting. (And the wording, using @RequiredArgsConstructor so much is making me forget to distinguish between final fields and constructor arguments sometimes.)
I guess I could add ElementType.PARAMETER to the annotation target and tell lombok to copy it to the generated constructor. Still, it would be great if it worked out-of-the-box.
I'll be happy to try to make a PR, if you think there's a chance the feature would get accepted.

**Comment by sbrannen** [comment]
> However, when I tried to use it, I hit a problem and I'm not sure if (or why) it is supposed to work like this. For some reason, non-static fields are only checked for extensions _after_ the test instance is created

The reason it works like that is that I wanted to minimize code duplication in `org.junit.jupiter.engine.descriptor.ExtensionUtils.registerExtensionsFromFields(...)` and reuse that method as much as possible. From a design perspective, we thought we had covered all reasonable use cases with that logic. In other words, we did not consider the specific use case you tried.

>  I'll be happy to try to make a PR, if you think there's a chance the feature would get accepted.

Thanks for offering, but please hold off on submitting a PR.

We will discuss this during our next team call and decide where to go from there.

**Comment by marcphilipp** [comment]
**Team decision:** Let's explore adding support for registering `@ExtendWith` extensions on fields as early as possible so they can be used for `ParameterResolver`, `TestInstancePostProcessor`, and other similar extensions.

**Comment by sbrannen** [comment]
> I guess I could add ElementType.PARAMETER to the annotation target and tell lombok to copy it to the generated constructor. Still, it would be great if it worked out-of-the-box.

Well, if you can use records, Java already does that for you.

For example, the following works out of the box.

```java
public record TestClass(@TestAnnotation String someField) {

	@Test
	void someTestMethod() {
		assertThat(this.someField).isEqualTo("abc");
	}
}

@Target(ElementType.PARAMETER)
@Retention(RetentionPolicy.RUNTIME)
@ExtendWith(TestAnnotationExtension.class)
@interface TestAnnotation {
}
```
